lib/wibox/widget/template.lua

```diff
@@ -0,0 +1,90 @@
+---------------------------------------------------------------------------
+-- An abstract widget that handles a preset of concrete widget.
+--
+-- The `wibox.widget.template` widget is an abstraction layer that contains a
+-- concrete widget definition. The template widget can be used to build widgets
+-- that the user can customize at their will, thanks to the template mechanism.
+--
+-- @author Aire-One
+-- @copyright 2021 Aire-One <aireone@aireone.xyz>
+--
+-- @classmod wibox.widget.template
+-- @supermodule wibox.widget.base
+---------------------------------------------------------------------------
+
+local wbase = require("wibox.widget.base")
+local gtable = require("gears.table")
+local gtimer = require("gears.timer")
+
+local default_update_callback = function()
+    return nil
+end
+
+local default_template_widget = wbase.empty_widget()
+
+local template = {
+    mt = {},
+    queued_updates = {}
+}
+
+function template:_do_update_now(args)
+    self:update_callback(args)
+    template.queued_updates[self] = false
+end
+
+--- Update the widget.
+-- This function will call the `update_callback` function at the end of the
+-- current GLib event loop. Updates are batched by event loop, it means that the
+-- widget can only be update once by event loop. If the `template:update` method
+-- is called multiple times during the same GLib event loop, only the first call
+-- will be run.
```

Comment:
I think it makes more sense to only run the _latest_ update request.

Use case:
It's very likely that this `:update()` method will be called from `property::` signal handlers. But those are executed synchronously.

So something like
```lua
mywidget.foo = "foo"
mywidget.bar = "bar"
```
would translate to
```lua
mywidget.foo = "foo"
  -- queue `:update()`
mywidget.bar = "bar"
  -- ignore

-- run `:update()` with `{ foo = "foo" }`
```
Whereas the desired outcome would be
```lua
mywidget.foo = "foo"
  -- queue `:update()`
mywidget.bar = "bar"
  -- discard previous, queue new `:update()`

-- run `:update()` with `{ foo = "foo", bar = "bar" }`
```

I don't think there is an easy way to discard a specific handler from delayed calls.
But another way to implement this would be using table references for `args` so that calling `:update()` always updates `<some_table>` and `:_do_update_now()` holds a reference to that same table.

Reply:
I have updated the way the update callback arguments are managed. Now they are saved to an attribute to be easily passed to the user defined callback.